[memory profiling] Move Context object to c10

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96281
* #95753
* #95541
* #95357
* __->__ #96280
* #95356

Minor refactor so that follow up PR can have objects that meet the GatheredContext
inferface without having to depend on CUDA.

zdevito committed:
[memory profiling] Move Context object to c10

Minor refactor so that follow up PR can have objects that meet the GatheredContext
inferface without having to depend on CUDA.

[ghstack-poisoned]

Skylion007 commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -152,15 +148,15 @@ struct TraceEntry {
+      std::shared_ptr<GatheredContext> context = nullptr)
       : action_(action),
         addr_(addr),
         context_(context),
```
```suggestion
        context_(std::move(context)),
```
No reason for an unnecessary copy here.

zdevito committed:
Update c10/cuda/CUDACachingAllocator.h

Co-authored-by: Aaron Gokaslan <aaronGokaslan@gmail.com>

zdevito committed:
Update on "[memory profiling] Move Context object to c10"


Minor refactor so that follow up PR can have objects that meet the GatheredContext
inferface without having to depend on CUDA.

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] Move Context object to c10"


Minor refactor so that follow up PR can have objects that meet the GatheredContext
inferface without having to depend on CUDA.

[ghstack-poisoned]
